java-runtime-builder-app/src/main/java/com/google/cloud/runtimes/builder/buildsteps/PrebuiltRuntimeImageBuildStep.java

```diff
@@ -54,7 +63,31 @@ protected String getArtifact(BuildContext buildContext) throws BuildStepExceptio
     } else if (artifacts.size() > 1) {
       throw new TooManyArtifactsException(artifacts);
     } else {
-      return buildContext.getWorkspaceDir().relativize(artifacts.get(0)).toString();
+      Path artifact = artifacts.get(0);
+      logger.debug("Found Java artifact {}", artifact);
+      return artifact;
     }
   }
+
+  /*
+   * Searches non-recursively for deployable artifacts in the given directory. Potential artifacts
+   * include:
+   * - files or directories ending in .jar or .war
+   * - the current directory itself, if it is an exploded war
+   */
+  private List<Path> findArtifacts(Path searchDir) throws IOException {
+    List<Path> artifacts = Files.list(searchDir)
+        .filter(path -> {
+          String extension = com.google.common.io.Files.getFileExtension(path.toString());
+          return extension.equalsIgnoreCase("war") || extension.equalsIgnoreCase("jar");
+        })
+        .collect(Collectors.toList());
+
+    // If the search directory contains a compat deployment descriptor, it is a valid artifact
+    if (Files.exists(searchDir.resolve("WEB-INF").resolve("appengine-web.xml"))) {
+      artifacts.add(searchDir);
```

Comment:
Maybe in this scenario, treating the actual path to the file `appengine-web.xml` as the artifact will make things a bit cleaner?